### DISCUSSION TITLE
$persist plugin doesn't work in iframe in incognito mode

### DISCUSSION THREAD

**Discussion by geouser** [alpinejs/alpine#3710] [7. Bugs (archived)]
Hi there!

Today I found an issue with the $persist plugin. When the page that uses AlpineJS is embedded on another website via iframe, and then you open it in incognito mode in your browser, with cookies disabled - it won't work. The $persist plugin will trigger an error, because `localStorage` is not defined on the `window` object. The issue was also mentioned here #3451 

![image](https://github.com/alpinejs/alpine/assets/13234430/ff816118-76e3-452b-9378-45f723cd6d24)

Here is a [fiddle](https://jsfiddle.net/d2ya3Ltr/) where this issue can be reproduced. Please open it in incognito mode in Chrome.

I understand that this is a very specific use case.

Thanks!

**Comment by ekwoka** [comment]
Well, nothing can be done about that.

you don't have localStorage there. So if you have that use case, you should be providing an alternative storage when localStorage isn't available.

Also, that doesn't say it's not defined. It says access is denied.

**Reply by geouser** [reply]
@ekwoka Yes, that is clear that localStorage can't be used, because it is disabled by the user (intentionally or not). The point here is that this error breaks the components, and it doesn't work.

Ideally, if the localStorage is not available, the plugin should just return the raw value that was passed to it.

> So if you have that use case, you should be providing an alternative storage when localStorage isn't available.

Alternative storage does not help in this case, because the error happens during the plugin initialization, and before the alternative storage is set with `.using()` method

> Also, that doesn't say it's not defined. It says access is denied.

You are right here. The access to localStorage is denied

**Comment by SimoTod** [comment]
@geouser the plugin is open source. You can probably send a PR to fix it with some try-catches on the lines reading/writing from/to the storage and it will guarantee the component works without the persisting functionality for those edge cases.

**Reply by ekwoka** [reply]
Seems an explicit check would be best, but I don't know if there is a way to check if it's accessible here....since it's not undefined...but just access denied...

That's a strange one.

**Reply by geouser** [reply]
This is what eventually I ended up with

```javascript
let storageSupported = false; 

try { 
  storageSupported = (window.localStorage && true) 
} catch (e) { }

if (!storageSupported) {
  Object.defineProperty(window, 'localStorage', {value: {
    _data       : {},
    setItem     : function(id, val) { return this._data[id] = String(val); },
    getItem     : function(id) { return this._data.hasOwnProperty(id) ? this._data[id] : undefined; },
    removeItem  : function(id) { return delete this._data[id]; },
    clear       : function() { return this._data = {}; }
  } });
}
```
